pkg/agent/supportbundlecollection/support_bundle_controller.go

```diff
@@ -0,0 +1,414 @@
+	"github.com/spf13/afero"
+	"golang.org/x/crypto/ssh"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/exec"
+
+	"antrea.io/antrea/pkg/agent"
+	agentquerier "antrea.io/antrea/pkg/agent/querier"
+	"antrea.io/antrea/pkg/apis/controlplane"
+	cpv1b2 "antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+	"antrea.io/antrea/pkg/ovs/ovsctl"
+	"antrea.io/antrea/pkg/querier"
+	"antrea.io/antrea/pkg/support"
+	"antrea.io/antrea/pkg/util/compress"
+)
+
+type ProtocolType string
+
+const (
+	sftpProtocol ProtocolType = "sftp"
+
+	controllerName = "SupportBundleCollectionController"
+
+	uploadToFileServerTries      = 5
+	uploadToFileServerRetryDelay = 5 * time.Second
+)
+
+var (
+	emptyWatch      = watch.NewEmptyWatch()
+	defaultFS       = afero.NewOsFs()
+	defaultExecutor = exec.New()
+	// Declared as variable for testing.
+	newAgentDumper = support.NewAgentDumper
+)
+
+type SupportBundleController struct {
+	nodeName                     string
+	supportBundleNodeType        controlplane.SupportBundleCollectionNodeType
+	namespace                    string
+	antreaClientGetter           agent.AntreaClientProvider
+	queue                        workqueue.Interface
+	supportBundleCollection      *cpv1b2.SupportBundleCollection
+	supportBundleCollectionMutex sync.RWMutex
+	ovsCtlClient                 ovsctl.OVSCtlClient
+	aq                           agentquerier.AgentQuerier
+	npq                          querier.AgentNetworkPolicyInfoQuerier
+	v4Enabled                    bool
+	v6Enabled                    bool
+	sftpUploader                 uploader
+}
+
+func NewSupportBundleController(nodeName string,
+	supportBundleNodeType controlplane.SupportBundleCollectionNodeType,
+	namespace string,
+	antreaClientGetter agent.AntreaClientProvider,
+	ovsCtlClient ovsctl.OVSCtlClient,
+	aq agentquerier.AgentQuerier,
+	npq querier.AgentNetworkPolicyInfoQuerier,
+	v4Enabled,
+	v6Enabled bool) *SupportBundleController {
+	c := &SupportBundleController{
+		nodeName:              nodeName,
+		supportBundleNodeType: supportBundleNodeType,
+		namespace:             namespace,
+		antreaClientGetter:    antreaClientGetter,
+		queue:                 workqueue.NewNamed("supportbundle"),
+		ovsCtlClient:          ovsCtlClient,
+		aq:                    aq,
+		npq:                   npq,
+		v4Enabled:             v4Enabled,
+		v6Enabled:             v6Enabled,
+		sftpUploader:          &sftpUploader{},
+	}
+	return c
+}
+
+func (c *SupportBundleController) watchSupportBundleCollections() {
+	klog.Info("Starting watch for SupportBundleCollections")
+	antreaClient, err := c.antreaClientGetter.GetAntreaClient()
+	if err != nil {
+		klog.ErrorS(err, "Failed to get antrea client")
+		return
+	}
+	options := metav1.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector("nodeName", c.nodeName).String(),
```

Comment:
Use `k8s.NamespacedName(c.namespace, c.nodeName)` as the ListOption with the "ExternalNode" case, which is expected after this PR https://github.com/antrea-io/antrea/pull/4401. Otherwise, ExternalNode would use an incorrect value when controller calculating the span.

Reply:
updated